src/sb-util.ts

```diff
@@ -1,3 +1,91 @@
-export function sb() {
-  console.log('sb-util')
-}
\ No newline at end of file
+import { Queryable, ScratchProjectOptions } from './abstracts';
+import { SpOptionsEmptyError, SpMultipleSourceOptionsError } from './errors';
+import AdmZip from 'adm-zip';
+import fs from 'fs';
+import fetch from 'isomorphic-fetch';
+
+enum ProjectSource {
+	FILE = 'file',
+	URI = 'uri',
+	CLOUD_ID = 'cloudId',
+	PROJECT_JSON = 'project.json',
+	MIT_PROJECT_URI = 'https://projects.scratch.mit.edu'
+}
+
+/*
+  The AssetFetcher is responsible for handing different
+  project sources and the asset files within or referred
+  to in a project.
+*/
+const AssetFetcher = {
+	parseSb3: (sb3File: string) => {
+		return new Promise((resolve, reject) => {
+			const sb3 = new AdmZip(sb3File);
+			const projectJSONString = sb3.readAsText(ProjectSource.PROJECT_JSON);
+			resolve(JSON.parse(projectJSONString));
+		})
+	},
+
+	parseJSON: (jsonFile: string) => {
+		return new Promise((resolve, reject) => {
+			fs.readFile(jsonFile, (err, data) => {
+				resolve(JSON.parse(data.toString()));
+			})
+		})
+	},
+
+	parseFromCloudID: (cloudId: number) => {
+		return new Promise(async (resolve, reject) => {
+			const data = await fetch(`${ProjectSource.MIT_PROJECT_URI}/${cloudId}`);
+			const projectJSON = await data.json();
+			resolve(projectJSON);
+		})
+	}
+}
+
+
+export class ScratchProject implements Queryable {
+	constructor(projectJSON: Object, assetFetcher?: any) {
+
+	}
+
+	query(selector?: String) {
+		return [];
+	}
+}
+
+/*
+	initialize() is a factory method that returns a
+	Promise to a ScratchProject while it does the
+	necessary I/O or HTTP requests needed to populate
+	the ScratchProject
+*/
+export const initialize = function(options: ScratchProjectOptions): Promise<ScratchProject> {
+	// Check project options before initializing a ScratchProject
+	if (!options || Object.keys(options).length === 0) { 
+		throw new SpOptionsEmptyError('Please provide one of the following options: file, uri, or cloudId.'); 
+	} if (Object.keys(options).length > 1) {
+		throw new SpMultipleSourceOptionsError('Multiple options found. Please supply only one of the following: file, uri, or cloudId');
+	}
+
+	let fileSource, parse;
+
+	if (ProjectSource.FILE in options) {
```

Comment:
Above you're validating `options` with `Object.keys(options)` which only includes _own properties_; using `in` here will check for properties all the way up the prototype chain.

Reply:
Object.getOwnProperty is synonymous with the prototype handling of Object.keys.